### PR TITLE
Additional HttpClientOptions Configurations

### LINKED ISSUE #21193: Support Additional Configurations in HttpClientOptions
Add support for additional, common configuration options in `HttpClientOptions`.

- [x] Connection pool size
- [x] Idle connection timeout

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes #21193 

This PR adds additional configuration options to `HttpClientOptions` allowing for easier configuration of an `HttpClient` implementation's connection management.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java [hunk 1]
```diff
@@ -191,13 +212,76 @@ public HttpClientOptions setReadTimeout(Duration readTimeout) {
...
      * @return The read timeout duration.
      */
     public Duration getReadTimeout() {
         return getTimeout(readTimeout);
     }
 
+    /**
+     * Sets the maximum connection pool size used by the underlying HTTP client.
+     * <p>
+     * By default the maximum connection pool size is determined by the underlying HTTP client.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java:227]: We should also note that this is not a global max. It is the max pool size for that specific client.

File: sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncClientProvider.java [hunk 2]
```diff
@@ -26,6 +29,15 @@ public HttpClient createInstance(HttpClientOptions clientOptions) {
                 .configuration(clientOptions.getConfiguration())
                 .writeTimeout(clientOptions.getWriteTimeout())
                 .readTimeout(clientOptions.getReadTimeout());
+
+            int maximumConnectionPoolSize = (clientOptions.getMaximumConnectionPoolSize() == 0)
+                ? 5 // By default OkHttp uses a maximum idle connection count of 5.
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncClientProvider.java:34]: It is a required parameter in the constructor, I'll check if they (OkHttp) have it as a constant anywhere.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java:227]: I'll update this

File: sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java [hunk 3]
```diff
@@ -191,13 +212,76 @@ public HttpClientOptions setReadTimeout(Duration readTimeout) {
...
+        this.connectionIdleTimeout = connectionIdleTimeout;
+        return this;
+    }
+
+    /**
+     * Gets the duration of time before an idle connection is closed.
+     * <p>
+     * By default the connection idle timeout is 60 seconds.
+     *
+     * @return The connection idel timeout duration.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java:279]:
```suggestion
     * @return The connection idle timeout duration.
```

File: sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java [hunk 4]
```diff
@@ -191,13 +212,76 @@ public HttpClientOptions setReadTimeout(Duration readTimeout) {
...
+    }
+
+    /**
+     * Gets the maximum connection pool size used by the underlying HTTP client.
+     * <p>
+     * By default the maximum connection pool size is determined by the underlying HTTP client.
+     *
+     * @return The maximum connection pool size.
+     */
+    public int getMaximumConnectionPoolSize() {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java:250]: Should this be `Integer` to return null if the user doesn't set the pool size?

File: sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java [hunk 5]
```diff
@@ -191,13 +212,76 @@ public HttpClientOptions setReadTimeout(Duration readTimeout) {
...
+    /**
+     * Sets the maximum connection pool size used by the underlying HTTP client.
+     * <p>
+     * By default the maximum connection pool size is determined by the underlying HTTP client.
+     *
+     * @param maximumConnectionPoolSize The maximum connection pool size.
+     * @return The updated HttpClientOptions object.
+     * @throws IllegalArgumentException If {@code maximumConnectionPoolSize} is less than {@code 1}.
+     */
+    public HttpClientOptions setMaximumConnectionPoolSize(int maximumConnectionPoolSize) {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java:233]: To allow for unsetting the pool size we should use `Integer` and allow null values. If not null, we should check that it's greater than 0.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java [hunk 6]
```diff
@@ -191,13 +212,94 @@ public HttpClientOptions setReadTimeout(Duration readTimeout) {
...
+     * Modifying the maximum connection pool size may have effects on the performance of an application. Increasing the
+     * maximum connection pool will result in more connections being available for an application but may result in
+     * more contention for network resources. It is recommended to perform performance analysis on different maximum
+     * connection pool sizes to find the right configuration for an application.
+     * <p>
+     * This maximum connection pool size is not a global configuration but an instance level configuration for each
+     * {@link HttpClient} created using this {@link HttpClientOptions}.
+     * <p>
+     * By default the maximum connection pool size is determined by the underlying HTTP client. Setting the maximum
+     * connection pool size resets the configuration to use the default determined by the underlying HTTP client.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java:264]:
```suggestion
     * connection pool size to {@code null} resets the configuration to use the default determined by the underlying HTTP client.
```

File: sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java [hunk 7]
```diff
@@ -191,13 +212,94 @@ public HttpClientOptions setReadTimeout(Duration readTimeout) {
...
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/HttpClientOptions.java:236]:
```suggestion
     * connection pool size to {@code null} resets the configuration to use the default determined by the underlying HTTP client.
```

**Review by srnagar — APPROVED** [review]

**Pull request merged**